lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -1385,6 +1386,110 @@ void TypeChecker::diagnoseDuplicateCaptureVars(CaptureListExpr *expr) {
+          Diag<bool> fixitID) -> InFlightDiagnostic {
+    bool isModifier = attr->isDeclModifier();
+
+    diagnoseForAccessNote(attr, VD, diagID, notes.reason, isModifier,
+                          attr->getAttrName(), VD->getDescriptiveKind());
+    return diagnoseForAccessNote(attr, VD, fixitID, isModifier);
+  };
+
+  if (*expected) {
+    attr = willCreate();
+    VD->getAttrs().add(attr);
+
+    SmallString<64> attrString;
+    llvm::raw_svector_ostream os(attrString);
+    attr->print(os, VD);
+
+    diagnoseChangeByAccessNote(diag::attr_added_by_access_note,
+                               diag::fixit_attr_added_by_access_note)
+      .fixItInsert(VD->getAttributeInsertionLoc(attr->isDeclModifier()),
+                   attrString);
+  } else {
+    VD->getAttrs().removeAttribute(attr);
+    diagnoseChangeByAccessNote(diag::attr_removed_by_access_note,
+                               diag::fixit_attr_removed_by_access_note)
+      .fixItRemove(attr->getRangeWithAt());
+  }
+}
+
+static void applyAccessNote(ValueDecl *VD, const AccessNote &note,
+                            const AccessNotes &notes) {
+  ASTContext &ctx = VD->getASTContext();
+
+  addOrRemoveAttr<ObjCAttr>(VD, notes, note.ObjC, [&]() {
+    return ObjCAttr::create(ctx, note.ObjCName, false);
+  });
+
+  addOrRemoveAttr<DynamicAttr>(VD, notes, note.Dynamic, [&]{
+    return new (ctx) DynamicAttr(true);
+  });
+
+  if (note.ObjCName) {
+    auto attr = VD->getAttrs().getAttribute<ObjCAttr>();
+    assert(attr && "ObjCName set, but ObjCAttr not true or did not apply???");
+
+    if (!attr->hasName()) {
+      attr->setName(*note.ObjCName, true);
+      diagnoseForAccessNote(attr, VD,
+                            diag::attr_objc_name_changed_by_access_note,
+                            notes.reason, VD->getDescriptiveKind(),
+                            *note.ObjCName);
+
+      SmallString<64> newNameString;
+      llvm::raw_svector_ostream os(newNameString);
+      os << "(";
+      os << *note.ObjCName;
+      os << ")";
+
+      auto note = diagnoseForAccessNote(attr, VD,
+                            diag::fixit_attr_objc_name_changed_by_access_note);
+
+      if (attr->getLParenLoc().isValid())
+        note.fixItReplace({ attr->getLParenLoc(), attr->getRParenLoc() },
+                          newNameString);
+      else
+        note.fixItInsertAfter(attr->getLocation(), newNameString);
+    }
+    else if (attr->getName() != *note.ObjCName) {
+      diagnoseForAccessNote(attr, VD,
+                            diag::attr_objc_name_conflicts_with_access_note,
+                            notes.reason, VD->getDescriptiveKind(),
+                            *note.ObjCName, attr->getName());
+    }
+  }
+}
+
+evaluator::SideEffect
+ApplyAccessNoteRequest::evaluate(Evaluator &evaluator, ValueDecl *VD) const {
+  AccessNotes &notes = VD->getModuleContext()->getAccessNotes();
+  if (auto note = notes.lookup(VD))
```

Comment:
Applying access notes lazily like this is mostly great, except that it becomes impossible to diagnose that an access note doesn't match anything. It might be better to walk through the access notes after parsing, looking up the declarations matching their names and applying them ahead of time.

Reply:
This needs more thought, so I'm adding a FIXME to the test and otherwise leaving it for later.